hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/batching/PartitionAwareCallableBatchingRunnable.java

```diff
@@ -0,0 +1,130 @@
+public class PartitionAwareCallableBatchingRunnable implements Runnable {
+
+    private final PartitionAwareCallableFactory factory;
+    private final IPartitionService partitionService;
+    private final Address memberAddress;
+    private final int partitionThreadCount;
+
+    private final CopyOnWriteArrayList results;
+    private final AtomicInteger finished;
+    private final ResultFuture resultFuture;
+
+    /**
+     * @param nodeEngine nodeEngine
+     * @param factory    factory of PartitionAwareCallable to be run on the partition thread.
+     */
+    public PartitionAwareCallableBatchingRunnable(NodeEngine nodeEngine, PartitionAwareCallableFactory factory) {
+        this.factory = factory;
+        this.partitionService = nodeEngine.getPartitionService();
+        this.memberAddress = nodeEngine.getThisAddress();
+        this.partitionThreadCount = ((OperationServiceImpl) nodeEngine.getOperationService()).getPartitionThreadCount();
+
+        this.finished = new AtomicInteger(0);
+        this.resultFuture = new ResultFuture(nodeEngine, nodeEngine.getLogger(getClass()));
+        this.results = new CopyOnWriteArrayList();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void run() {
+        if (!(Thread.currentThread() instanceof PartitionOperationThread)) {
+            throw new IllegalArgumentException("The PartitionAwareCallableBatchingRunnable has to be run on a partition-thread");
+        }
+        PartitionOperationThread currentThread = ((PartitionOperationThread) Thread.currentThread());
+        int partitionThreadId = currentThread.getThreadId();
+        try {
+            int currentPartitionId = partitionThreadId;
+            IPartition[] partitions = partitionService.getPartitions();
+            runSequentially(currentThread, currentPartitionId, partitions);
+        } finally {
+            int value = finished.incrementAndGet();
+            if (!resultFuture.isDone() && value == partitionThreadCount) {
+                resultFuture.setResult(results);
+            }
+        }
+    }
+
+    private void runSequentially(PartitionOperationThread currentThread, int currentPartitionId, IPartition[] partitions) {
+        while (currentPartitionId < partitions.length) {
+            final IPartition partition = partitions[currentPartitionId];
+            if (currentThread.isInterrupted()) {
+                resultFuture.cancel(true);
+                break;
+            }
+            PartitionAwareCallable task = factory.create();
+            if (memberAddress.equals(partition.getOwnerOrNull())) {
+                try {
+                    results.add(task.call(currentPartitionId));
+                } catch (Exception ex) {
+                    resultFuture.setResult(ex);
+                    break;
+                }
+            }
+            currentPartitionId += partitionThreadCount;
+        }
+    }
+
+    public ICompletableFuture getFuture() {
+        return resultFuture;
```

Comment:
It is better to call this field 'future'. A field already implies a handle to a result that will appear in some moment in time. Also the getter is called 'getFuture'; you don't want to have 2 names for the same thing.

Reply:
Agree, changed.